BugsnagNetworkRequestPlugin/BugsnagNetworkRequestPluginTests/BSGURLSessionTracingDelegateTests.m

```diff
@@ -0,0 +1,68 @@
+//
+//  BSGURLSessionTracingDelegateTests.m
+//  BugsnagNetworkRequestPlugin
+//
+//  Created by Nick Dowell on 22/09/2021.
+//
+
+#import "BSGURLSessionTracingDelegate.h"
+
+#import <XCTest/XCTest.h>
+
+@interface BSGURLSessionTracingDelegateTests : XCTestCase
+
+@end
+
+@implementation BSGURLSessionTracingDelegateTests
+
+- (void)testUrlParamsForQueryItems {
+#define TEST(url, expected) \
+XCTAssertEqualObjects([BSGURLSessionTracingDelegate urlParamsForQueryItems:[NSURLComponents componentsWithString:url].queryItems], expected)
+    
+    TEST(@"http://example.com", nil);
+    
+    TEST(@"http://example.com?", @{});
+    
+    TEST(@"http://example.com?foo=bar", @{@"foo": @"bar"});
+    
+    TEST(@"http://example.com?foo=bar&bar=baz", (@{@"foo": @"bar", @"bar": @"baz"}));
+    
+    // Multiple query items with the same name should be represented as arrays.
+    TEST(@"http://example.com?foo=bar&foo=baz", (@{@"foo": @[@"bar", @"baz"]}));
+    
+    // Query items with no value should be represented as empty string.
+    TEST(@"http://example.com?foo=bar&foo=baz&foo=&sort=name", (@{@"foo": @[@"bar", @"baz", @""], @"sort": @"name"}));
```

Comment:
This is missing a few cases.

1. without an equals sign, like

   ```objc
   TEST(@"http://example.com?bar", @{@"bar":@""});
   ```
   where the query item's value would be nil

2. unterminated `&`

   ```objc
   TEST(@"http://example.com?foo=bar&", @{@"foo": @"bar", @"":@""});
   ```

3. leading `=`

   ```objc
   TEST(@"http://example.com?=bar", @{@"": @"bar"});
   ```

Reply:
Added these cases in https://github.com/bugsnag/bugsnag-cocoa/pull/1193/commits/5883a9e1d975430eaa90e7e6ec6ea2fb2608a0fc, though the results don't quite match your expectations 🙃

Reply:
I'm not sure I follow, previously you mentioned: 

> Name-value pairs with no equals sign are represented as an NSURLQueryItem with `item.value = @""`

Though if that was the case, [this test](
https://github.com/bugsnag/bugsnag-cocoa/commit/5883a9e1d975430eaa90e7e6ec6ea2fb2608a0fc#diff-6fad19de5bea41e743cf47d78cfe028632b8e00fe3a8f675e6819704ac1c16f7R36) would have a value?

```objc
TEST(@"http://example.com?foo", @{});
```

It looks like the value is nil, which is resulting in the key being "unset" [here](https://github.com/bugsnag/bugsnag-cocoa/pull/1193/files#diff-9184f3f6165422fee0871d7b405793441bee81d6be485fa7101129696675094bR59), so you can't tell the original query params contained key `foo`.

Reply:
My mistake, I had not realized we would get an NSURLQueryItem with a nil value... should have read the documentation!

```C
// - note: If a NSURLQueryItem name-value pair is empty (i.e. the query string starts with '&', ends
// with '&', or has "&&" within it), you get a NSURLQueryItem with a zero-length name and a nil value.
// If a NSURLQueryItem name-value pair has nothing before the equals sign, you get a zero-length name.
// If a NSURLQueryItem name-value pair has nothing after the equals sign, you get a zero-length value.
// If a NSURLQueryItem name-value pair has no equals sign, the NSURLQueryItem name-value pair string
// is the name and you get a nil value.
```

Reply:
Ok https://github.com/bugsnag/bugsnag-cocoa/pull/1193/commits/3833e06503de156e0b9846deeb9139385162370f should hopefully resolve this